cpp/src/strings/search/contains_multiple.cu

```diff
@@ -0,0 +1,315 @@
+ * in the d_indices array:
+ * - tgt1_idx = d_indices[map_idx]   = 3 --> d_targets[3] == 'hi'
+ * - tgt2_idx = d_indices[map_idx+1] = 1 --> d_targets[1] == 'hello'
+ * The logic now only needs to check for either of these 2 targets.
+ *
+ * This kernel works in either row-per-string or warp-per-string depending
+ * on the template parameter. If tile_size==1, then it executes as a
+ * row-per-string. If tile_size=32, the it executes as a warp-per-string.
+ * No other options are supported for now.
+ *
+ * @tparam tile_size Number of threads per string
+ * @param d_strings Input strings
+ * @param d_targets Target strings to search within input strings
+ * @param d_first_bytes Sorted, unique list of first bytes of the target strings
+ * @param d_indices Indices to map sorted d_first_bytes to d_targets
+ * @param d_offsets Offsets to map d_indices to d_targets
+ * @param unique_count Number of unique values in d_first_bytes (and d_offsets)
+ * @param working_memory Global memory to use if shared-memory is too small
+ * @param d_results Bool results for each target within each string row
+ */
+template <cudf::thread_index_type tile_size>
+CUDF_KERNEL void multi_contains_kernel(column_device_view const d_strings,
+                                       column_device_view const d_targets,
+                                       u_char const* d_first_bytes,
+                                       size_type const* d_indices,
+                                       size_type const* d_offsets,
+                                       size_type unique_count,
+                                       bool* working_memory,
+                                       cudf::device_span<bool*> d_results)
+{
+  auto const idx     = cudf::detail::grid_1d::global_thread_id();
+  auto const str_idx = idx / tile_size;
+  if (str_idx >= d_strings.size()) { return; }
+  if (d_strings.is_null(str_idx)) { return; }
+
+  // get the string for this tile
+  auto const d_str = d_strings.element<string_view>(str_idx);
+
+  auto const lane_idx    = idx % tile_size;
+  auto const tile_idx    = static_cast<cudf::thread_index_type>(threadIdx.x) / tile_size;
+  auto const num_targets = d_targets.size();
+
+  // size of shared_bools = num_targets * block_size
+  // each thread uses num_targets bools
+  extern __shared__ bool shared_bools[];
+  // bools for the current string
+  auto bools = working_memory == nullptr ? (shared_bools + (tile_idx * tile_size * num_targets))
+                                         : (working_memory + (str_idx * tile_size * num_targets));
+
+  // initialize result: set true if target is empty, false otherwise
+  for (auto target_idx = lane_idx; target_idx < num_targets; target_idx += tile_size) {
+    auto const d_target = d_targets.element<string_view>(target_idx);
+    if constexpr (tile_size == 1) {
+      d_results[target_idx][str_idx] = d_target.empty();
+    } else {
+      auto const begin = bools + (target_idx * tile_size);
```

Comment:
When using shared memory and assume num_targets = 32, `block_size * targets.size()` = 256 * 32 = 8K shared memory. But the max offset of `target_idx * tile_size` is max_target 32 * tile_size 32 = 1K. Did not use all the shared memory.